src/output/plugins/visualization/SoundAnalysis.cxx

```diff
@@ -0,0 +1,394 @@
+  out(fftwf_alloc_complex(out_samples * num_channels), fftwf_free),
+  bass_mids_trebs(new float[3 * num_channels])
+{
+	if (num_samples > INT_MAX) {
+		throw FmtInvalidArgument("num_samples({}) may not be larger than {}",
+								 num_samples, INT_MAX);
+	}
+	
+	int n[] = { (int)num_samples };
+
+	/* The input is assumed to be interleaved; this seems convenient from the
+	 * perspective of how it's stored from the AudioOutput... tho if we need an
+	 * additional copy to convert it to `float`, we'd have the opportunity to
+	 * re-arrange it. */
+
+	int dist = num_samples;
+
+	/* Per the FFTW docs:
+	 * 
+	 * "`rank` is the rank of the transform (it should be the size of the array
+	 * `*n`) we use the term rank to denote the number of independent indices in
+	 * an array. For example, we say that a 2d transform has rank 2, a 3d
+	 * transform has rank 3, and so on."
+	 * 
+	 * This is always 1, for us.
+	 * 
+	 * layout of `in`:
+	 * 
+	 * | 0 ... num_samples-1 | num_samples ... 2*num_samples-1 | 2*num_samples ...
+	 * | data for chan 0	 | data for chan 1				   | data for chan 2 */
+ 
+	/* `howmany` is the number of transforms to compute. The resulting plan
+	 * computes `howmany` transforms, where the input of the k-th transform is at
+	 * location in+k*idist (in C pointer arithmetic), and its output is at
+	 * location out+k*odist. */
+
+	int odist = (num_samples / 2) + 1;
+
+	plan = fftwf_plan_many_dft_r2c(1,			  // rank of the input array-- we have one-dimensional arrays
+								   n,			  // the number of elements in each array
+								   num_channels,  // one array for each channel
+								   in.get(),	  // input buffer-- need to copy samples here before executing
+								   NULL,
+								   1,			  // input stride
+								   dist,		  // distance between successive arrays (indexes, not bytes)
+								   out.get(),	  // output buffer-- overwritten on each execution
+								   NULL,
+								   1,			  // output stride
+								   odist,		  // distance between successive arrays (indexes, not bytes)
+								   FFTW_ESTIMATE);// should probably be zero (to select a more exhaustive
+												  // search), but out of an abundance of caution, tell
+												  // FFTW to pick a plan quickly
+	if (NULL == plan) {
+		throw FmtRuntimeError("Failed to generate an FFTW plan: num_samp={},num_chan={}",
+							  num_samples, num_channels);
+	}
+
+	freq_lo = params.GetLoCutoff();
+
+	float samples_per_sec = (float) audio_format.GetSampleRate();
+	float ns = (float) num_samples;
+	// The highest frequency we can represent will be
+	float max_freq = (ns - 1.0f) * samples_per_sec / ns;
+	if (max_freq < params.GetHiCutoff()) {
+		FmtWarning(d_sound_analysis, "Clamping max frequency from {} to {}", freq_hi, max_freq);
+		freq_hi = max_freq;
+	} else {
+		freq_hi = params.GetHiCutoff();
+	}
+
+	idx_lo = (size_t) floor(freq_lo * (float) num_samples / samples_per_sec );
+	idx_hi = (size_t) ceil(freq_hi * (float)num_samples / samples_per_sec);
```

Comment:
Why use `double` functions when the parameter is `float`? This adds an implicit cast and uses a function that may be slower than necessary.

Reply:
fixed.